src/Stratis.Bitcoin.Features.BlockStore/BlockStoreLoop.cs

```diff
@@ -264,6 +264,9 @@ private void StartLoop()
                 var result = await this.stepChain.ExecuteAsync(nextChainedBlock, disposeMode, cancellationToken).ConfigureAwait(false);
                 if (result == StepResult.Stop)
                     break;
+
+                if (result == StepResult.Continue)
```

Comment:
Ah interesting you covered all scenarios?

Reply:
@noescape00 A little further up we check to see if the store's tip is higher than consensus. If it is we break out of this loop. (Line 254)... I.e. we are waiting for the store to catch up to consensus before we download more. Doesn't this already cover your scenario?

Reply:
During the IBD 'contunue' is returned only from ProcessPendingStorageStep.ProcessWhenInIBDAsync- here it returns continue no matter how many blocks were pushed to the block repository (those crazy iterations numbers are encountered when no blocks were pushed). And for some reason that 'tip is higher than consensus' check is passed.

I've tested it a few times - this behavior is encountered only with low peers count or when peers are slow (otherwise we get as many 'continue' as 30% of all blocks downloaded).

In a normal scenario this fix doesn't reduce the CPU usage. 

Hmm. This actually might be connected to the race condition that I was trying to find for some time https://github.com/stratisproject/StratisBitcoinFullNode/issues/599 so I'm not sure now if the delay is the right fix for the problem. I probably should investigate it more in depth to find out if that 'continue' is justified at all times (because it looks like in some cases it doesn't).

Reply:
@noescape00 yes, Continue is only returned from the ProcessPendingStorage step. But I see its returned from both ProcessWhenInIBDAsync and ProcessWhenNotInIBDAsync.

I have a suspicion the high iteration count comes from us not adding a delay when the pending storage count is less than the processing threshold:

```
        private async Task<StepResult> ProcessWhenInIBDAsync(ProcessPendingStorageContext context)
        {
            this.logger.LogTrace("({0}:'{1}',{2}.{3}:{4})", nameof(context.NextChainedBlock), context.NextChainedBlock, nameof(this.BlockStoreLoop.PendingStorage), nameof(this.BlockStoreLoop.PendingStorage.Count), this.BlockStoreLoop.PendingStorage.Count);

            if (this.BlockStoreLoop.PendingStorage.Count < BlockStoreLoop.PendingStorageBatchThreshold)
                return StepResult.Continue;
```

Thoughts?

Reply:
Yes, that might be on option. 
But I'll also have to check how new blocks downloading is scheduled and the probability of that delay slowing down IBD overall.  

Ideally I'd switch from the async loop + inner while() with some kind of callback system like I did in blockStore signaled but taking in count the complexity of block pulling process that might be a task for later. 

Right now the plan is to tweak my server node to only connect to a local node and slow down the uploading process. I think that way I will be able to reproduce consistently what is causing lot's of 'continue' returns and start fixing that. Maybe that setup will also let me catch the race condition.

Reply:
Yeah, maybe post the logs here, so that I can also see what is being produced when lots of those "continues" happen ;)